### PR TITLE
fix(connector): [AUTHORIZEDOTNET] Fix status mapping

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- The status mapping for connector `Authorizdotnet` is fixed.
- The `ref_id` field being passed in connector contains `connector_request_reference_id`. The max number of characters allowed in ref_id is 20 and hence a check has been added to only pass ref_id if connector_request_reference_id contains `less than or equal to` 20 characters.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

https://github.com/juspay/hyperswitch/issues/3846
https://github.com/juspay/hyperswitch/issues/3847

## How did you test it?

Following flows should be tested for connector Authorizdotnet:

1. Authorization(Auto Capture):
`{
    "payment_id": "abcd1",
    "amount": 800,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "customer_id": "StripeCustomer",
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        }
    }
}`
NOTE: Try passing the above curl without payment_id also.
Response: `{
    "payment_id": "abcd1",
    "merchant_id": "merchant_1708944265",
    "status": "succeeded",
    "amount": 800,
    "net_amount": 800,
    "amount_capturable": 0,
    "amount_received": 800,
    "connector": "authorizedotnet",
    "client_secret": "abcd1_secret_eiryjwdDZgV6IFO8TeEu",
    "created": "2024-02-27T13:19:55.615Z",
    "currency": "USD",
    "customer_id": "StripeCustomer",
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": "42424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe"
        }
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer",
        "created_at": 1709039995,
        "expires": 1709043595,
        "secret": "epk_d639baf128604e01adc764d9d48437c6"
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "80015300949",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "80015300949",
    "payment_link": null,
    "profile_id": "pro_9LdgkmKiHJ2dR3oUn5Ow",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_HDQdFPQctm5R737tT7qf",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "expires_on": "2024-02-27T13:34:55.615Z",
    "fingerprint": null
}`

2. Authorization(Manual Capture):
`{
    "payment_id": "1ss",
    "amount": 800,
    "currency": "USD",
    "confirm": true,
    "capture_method": "manual",
    "customer_id": "StripeCustomer",
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        }
    }
}`
NOTE: Try passing the above curl without payment_id also.
Response: `{
    "payment_id": "1ss",
    "merchant_id": "merchant_1708944265",
    "status": "requires_capture",
    "amount": 1200,
    "net_amount": 1200,
    "amount_capturable": 1200,
    "amount_received": null,
    "connector": "authorizedotnet",
    "client_secret": "1ss_secret_4DnENIEpIUKKKnoKKElP",
    "created": "2024-02-27T13:20:40.683Z",
    "currency": "USD",
    "customer_id": "StripeCustomer",
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": "42424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe"
        }
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer",
        "created_at": 1709040040,
        "expires": 1709043640,
        "secret": "epk_a94c6c4d1e594998a168081802130610"
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "80015300969",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "80015300969",
    "payment_link": null,
    "profile_id": "pro_9LdgkmKiHJ2dR3oUn5Ow",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_HDQdFPQctm5R737tT7qf",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "expires_on": "2024-02-27T13:35:40.683Z",
    "fingerprint": null
}`

3. Capture:
`curl --location 'http://localhost:8080/payments/1ss/capture' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data '{}'`
Response:  `{
    "payment_id": "1ss",
    "merchant_id": "merchant_1708944265",
    "status": "succeeded",
    "amount": 1200,
    "net_amount": 1200,
    "amount_capturable": 0,
    "amount_received": 1200,
    "connector": "authorizedotnet",
    "client_secret": "1ss_secret_4DnENIEpIUKKKnoKKElP",
    "created": "2024-02-27T13:20:40.683Z",
    "currency": "USD",
    "customer_id": "StripeCustomer",
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": "42424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe"
        }
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "80015300969",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "80015300969",
    "payment_link": null,
    "profile_id": "pro_9LdgkmKiHJ2dR3oUn5Ow",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_HDQdFPQctm5R737tT7qf",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "expires_on": "2024-02-27T13:35:40.683Z",
    "fingerprint": null
}`
4. Void:
`curl --location 'http://localhost:8080/payments/1s2/cancel' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: API_KEY_HERE' \
--data '{}'`
Response: `{
    "payment_id": "1s2",
    "merchant_id": "merchant_1708944265",
    "status": "cancelled",
    "amount": 120,
    "net_amount": 120,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": "authorizedotnet",
    "client_secret": "1s2_secret_f0Q0ROMla0buxvBmaqR9",
    "created": "2024-02-27T13:22:23.892Z",
    "currency": "USD",
    "customer_id": "StripeCustomer",
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "4242",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "424242",
            "card_extended_bin": "42424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe"
        }
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "credit",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": false,
    "connector_transaction_id": "80015301006",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "80015301006",
    "payment_link": null,
    "profile_id": "pro_9LdgkmKiHJ2dR3oUn5Ow",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_HDQdFPQctm5R737tT7qf",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "expires_on": "2024-02-27T13:37:23.892Z",
    "fingerprint": null
}`

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 1]
```diff
@@ -536,35 +549,212 @@ pub enum AuthorizedotnetVoidStatus {
 impl From<AuthorizedotnetVoidStatus> for enums::AttemptStatus {
     fn from(item: AuthorizedotnetVoidStatus) -> Self {
         match item {
-            AuthorizedotnetVoidStatus::Approved => Self::VoidInitiated,
-            AuthorizedotnetVoidStatus::Declined | AuthorizedotnetVoidStatus::Error => Self::Failure,
+            AuthorizedotnetVoidStatus::Approved => Self::Voided,
+            AuthorizedotnetVoidStatus::Declined | AuthorizedotnetVoidStatus::Error => {
+                Self::VoidFailed
+            }
```
Comment (ArjunKarthik) [inline crates/router/src/connector/authorizedotnet/transformers.rs:555]: HeldForReview should be mapped to VoidInitiated

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 2]
```diff
@@ -536,35 +549,212 @@ pub enum AuthorizedotnetVoidStatus {
...
             AuthorizedotnetPaymentsResponse,
-            T,
+            types::PaymentsAuthorizeData,
             types::PaymentsResponseData,
         >,
     ) -> Result<Self, Self::Error> {
         match &item.response.transaction_response {
             Some(TransactionResponse::AuthorizedotnetTransactionResponse(transaction_response)) => {
-                let status = enums::AttemptStatus::from(transaction_response.response_code.clone());
+                let status = enums::AttemptStatus::foreign_from((
```
Comment (ArjunKarthik) [inline crates/router/src/connector/authorizedotnet/transformers.rs:582]: Don't duplicate the try_from

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 3]
```diff
@@ -536,35 +549,212 @@ pub enum AuthorizedotnetVoidStatus {
...
+            AuthorizedotnetPaymentsResponse,
+            types::CompleteAuthorizeData,
+            types::PaymentsResponseData,
+        >,
+    ) -> Result<Self, Self::Error> {
+        match &item.response.transaction_response {
+            Some(TransactionResponse::AuthorizedotnetTransactionResponse(transaction_response)) => {
+                let status = enums::AttemptStatus::foreign_from((
+                    transaction_response.response_code.clone(),
+                    true,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/authorizedotnet/transformers.rs:670]: Please confirm this whether wallets support manual capture or not

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 1 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs:552]: Done

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 2 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs]: Done

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 3 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs]: I have added is_auto_capture for CompleteAuthorize also

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 4]
```diff
@@ -536,35 +546,45 @@ pub enum AuthorizedotnetVoidStatus {
...
-        >,
+        data: (
+            types::ResponseRouterData<
+                F,
+                AuthorizedotnetPaymentsResponse,
+                T,
+                types::PaymentsResponseData,
+            >,
+            bool,
+        ),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/authorizedotnet/transformers.rs:579]:
```suggestion
        (item, is_authorize): (
            types::ResponseRouterData<
                F,
                AuthorizedotnetPaymentsResponse,
                T,
                types::PaymentsResponseData,
            >,
            bool,
        ),
```

Comment (deepanshu-iiitu) [reply crates/router/src/connector/authorizedotnet/transformers.rs]: Done

**Review by ArjunKarthik — APPROVED** [review]

**Review by Sakilmostak — APPROVED** [review]

**Pull request merged**